src/triggers/newEVMEvent.ts

```diff
@@ -0,0 +1,141 @@
+import _ from 'lodash';
+
+import { ContractFilter } from '../clients/evm';
+import { ChainId } from '../types/chain';
+import { Clients } from '../types/processor';
+import { Trigger, Cursor, TriggerOutput } from '../types/trigger';
+
+const NUMBER_OF_CONFIRMATIONS = BigInt(12);
+
+const min = (a: bigint, b: bigint) => (a < b ? a : b);
+
+// This calculates which range of blocks to query. We want to avoid redoing
+// queries for contracts that are already up to date. We also want to avoid
+// querying the same block range more than one. The below algorithm achieves
+// This by finding the most stale contracts and catching them up until all contracts
+// are in sync at the same block, after which we can query them all together.
+const calculateRange = (cursor: ContractsEventsCursor, gap: string, limit: number) => {
+  const contracts = Object.keys(cursor);
+  const contractsByBlock = _.groupBy(contracts, contract => cursor[contract]);
+  const blocks = Object.keys(contractsByBlock).map(b => BigInt(b));
+  const sortedBlocks = blocks.sort();
+  const mostStaleBlock = sortedBlocks[0];
+  const mostStaleContracts = contractsByBlock[mostStaleBlock.toString()].slice(0, limit);
+  const rangeStart: bigint = mostStaleBlock + BigInt(1);
+
+  let rangeEnd: bigint = rangeStart + BigInt(gap);
+  if (sortedBlocks.length > 1) {
+    rangeEnd = min(rangeEnd, sortedBlocks[1]);
+  }
+
+  return { rangeStart, rangeEnd, mostStaleContracts };
+};
+
+type ContractsEventsCursor = {
+  [contractAddress: string]: string;
+};
+
+export type EVMEventCursorArgs = {
+  id: string;
+  address: string;
+  startingBlock: string;
+};
+
+export const newEVMEvents: (
+  chainId: ChainId,
+  args: EVMEventCursorArgs[],
+  contractFilters: ContractFilter[],
+  gap?: string,
+  staleFilter?: (contractFilters: ContractFilter[], staleArgIds: string[]) => ContractFilter[],
+  updateCursor?: (rangeEnd: string, staleContractFilters: ContractFilter[]) => { [x: string]: string },
+  limit?: number
+) => Trigger<Cursor> = (
+  chainId: ChainId,
+  args: EVMEventCursorArgs[],
+  contractFilters: ContractFilter[],
+  gap: string = process.env.ETHEREUM_BLOCK_QUERY_GAP!,
+  staleFilter,
+  updateCursor,
+  limit = 5000
```

Comment:
add a limit to get the last {limit} arguments from the cursor, to prevent RPC queries with a huge number of OR conditions